### ISSUE TITLE
[Lab 4] Joins base

### ISSUE THREAD

**Issue by Ghoeb** [PUC-IIC3141/2017-1#9]
Se supone que SimpleDB trae implementado (aunque sea de manera primitiva) los joins, no?

Porque lo que hace cuando hago una query es siempre entregar el producto cruz de todas las tablas (esto sin incluir mi código) Pasa tanto en el `BasicQueryPlanner` como en el `HeuristicQueryPlanner`

No tengo que especificar de algun modo el campo por el que las tablas deben hacer join, no?  Como está implementado puede salir cualquier cosa:

Ejemplo de output:

```SQL
SQL> select n,x1 from X1

    x1     n
------------
     0     0
     1     1
     2     2
     3     3
     4     4

SQL> select n,x2 from X2

    x2     n
------------
     0     0
     1     1
     4     2
     9     3
    16     4

SQL> select n,x1,x2 from X1,X2


    x1    x2     n
------------------
     0     0     0
     0     1     0
     0     4     0
     0     9     0
     0    16     0
     1     0     1
     1     1     1
     1     4     1
     1     9     1
     1    16     1
     2     0     2
     2     1     2
     2     4     2
     2     9     2
     2    16     2
     3     0     3
     3     1     3
     3     4     3
     3     9     3
     3    16     3
     4     0     4
     4     1     4
     4     4     4
     4     9     4
     4    16     4

SQL> select n,x1,x2 from X1,X2 where x1 = x2

    x1    x2     n
------------------
     0     0     0
     1     1     1
     4     4     4
```

**Comment by cirojas** [comment]
> Se supone que SimpleDB trae implementado (aunque sea de manera primitiva) los joins, no?

Sí, se hace product (o multibuffer product) seguido de una selección, y si existe un índice en el atributo necesario, se puede hace un index join.

> Como está implementado puede salir cualquier cosa

SimpleDB tiene varias cosas incompletas, una de ellas es que no maneja bien el caso donde 2 tablas tienen un atributo con el mismo nombre, en tu ejemplo ambas tablas tienen el atributo n y eso confunde al motor. Por la forma en que está implementado, al juntar dos o más schemas se considera el atributo de solo una tabla y se ignorará el de las otras.

Puedes suponer que las consultas para testear cumplirán esta restricción y no existirán 2 tablas con nombres de atributos repetidos.

**Comment by Ghoeb** [comment]
Pero si no tienen atributos en comun, en base a que se hace el Join?

**Comment by cirojas** [comment]
Se hace un producto cruz sin ninguna selección posterior.

**Comment by crivero1** [comment]
De cierta forma, para la consulta que das corresponde que haga el producto cruz. En un motor convencional, la consulta que das más arriba daría error ya que no sabría de donde estas tomando el valor n. Cabe decir que el join en SQL debe ser explicito usando igualdad de atributos y no se asume directamente que es un "natural join". Por otro lado, como bien dice Carlos, simpleDB no tiene todo implementado y en este caso no maneja bien el uso de atributos comunes en dos tablas (tratando al final de hacer cualquier cosa).

**Comment by Ghoeb** [comment]
Perfecto. Está funcionando como corresponde entonces. Pero en ese caso, no entiendo el proposito del laboratorio, dado que al hacer producto cruz para todas las tablas el costo es el mismo independiente del orden en el que se haga... o estoy mal? Porque si todos los ordenes dan el mismo costo, como pretenden evaluar que efectivamente estemos encontrando el óptimo?

**Comment by cirojas** [comment]
No es lo mismo, porque después del producto cruz, si es que los 2 planes que se unen deben cumplir cierta condición, por ejemplo fld1=fld2 entonces viene inmediatamente una selección. Además dado el diseño de SimpleDB, si implementas correctamente la optimización de Selinger, debiese ser fácilmente extendible para incluir nuevos operadores de Join más eficientes.

También se me olvidó mencionar que está implementado el merge join (en el package  `simpledb.materialize` ), aunque nunca se hace uso de este. Si quieren pueden incluirlo en sus planes pero no es necesario (no lo he probado lo sufuciente para saber si funciona correctamente).

> como pretenden evaluar que efectivamente estemos encontrando el óptimo?

Como primera prueba pueden ver cual es la estimación usando `HeuristicQueryPlanner`, su solución siempre tiene que entregar estimaciones iguales o mejores. Además, si no modificaron nada en la estimación (`recordsOutput`) y usan correctamente la clase `TablePlanner`, deberían todos obtener siempre la misma estimación para una misma instancia de la BD.
